opentracing/src/main/java/opentracing/TraceContextDecoder.java

```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Decodes a trace context from binary or text formats.
+ *
+ * @see TraceContextEncoder
+ */
+// See https://github.com/opentracing/opentracing.github.io/issues/21 about folding codec together
+public interface TraceContextDecoder {
```

Comment:
>
>
>    1. Why have every en/de-coder responsible for the two kinds of
>    operations here?
>
> I don't recall the exact reason, but I think the reason for the split was
that not everyone is propagating context in text format (such as zipkin
does). Ex. in zipkin, propagation is via HTTP headers, so a map of string,
string works. the binary encoding isn't a feature we'd use.


>
>    1. Why is the EncodedTraceContext in the middle? Maybe straight from a
>    parameterized type T to aTraceContext implementation?
>
> There's probably an opportunity to link (or-repaste) rationale
documentation about the split of trace id vs other attributes. In following
other impls, they return a list of 2 collections each.

I agree that the encoded form could be more straight-forward if a type T.
Ex. I could use OkHttp's Headers object or a grpc Context directly that
way. If you feel interested enough, I could touch more on why I think
"straight-pipe" could be helpful for implementors.. want to raise an issue
here? https://github.com/opentracing/opentracing.github.io/issues